server/src/main/java/org/elasticsearch/index/engine/ThreadPoolMergeScheduler.java

```diff
@@ -0,0 +1,440 @@
+        }
+    }
+
+    /**
+     * Does the actual merge, by calling {@link org.apache.lucene.index.MergeScheduler.MergeSource#merge}
+     */
+    protected void doMerge(MergeSource mergeSource, MergePolicy.OneMerge merge) throws IOException {
+        mergeSource.merge(merge);
+    }
+
+    @Override
+    public Directory wrapForMerge(MergePolicy.OneMerge merge, Directory in) {
+        // Return a wrapped Directory which has rate-limited output.
+        // Note: the rate limiter is only per thread. So, if there are multiple merge threads running
+        // and throttling is required, each thread will be throttled independently.
+        // The implication of this, is that the total IO rate could be higher than the target rate.
+        RateLimiter rateLimiter = onGoingMergeRateLimiter.get();
+        return new FilterDirectory(in) {
+            @Override
+            public IndexOutput createOutput(String name, IOContext context) throws IOException {
+                ensureOpen();
+
+                // This Directory is only supposed to be used during merging,
+                // so all writes should have MERGE context, else there is a bug
+                // somewhere that is failing to pass down the right IOContext:
+                assert context.context() == IOContext.Context.MERGE : "got context=" + context.context();
+
+                return new RateLimitedIndexOutput(rateLimiter, in.createOutput(name, context));
+            }
+        };
+    }
+
+    final class MergeTask extends AbstractRunnable implements Comparable<MergeTask> {
+        private final String name;
+        private final SetOnce<Long> mergeStartTimeNS;
+        private final MergeSource mergeSource;
+        private final OnGoingMerge onGoingMerge;
+        private final MergeRateLimiter rateLimiter;
+        private final boolean supportsIOThrottling;
+
+        MergeTask(MergeSource mergeSource, MergePolicy.OneMerge merge, boolean supportsIOThrottling, String name) {
+            this.name = name;
+            this.mergeStartTimeNS = new SetOnce<>();
+            this.mergeSource = mergeSource;
+            this.onGoingMerge = new OnGoingMerge(merge);
+            this.rateLimiter = new MergeRateLimiter(merge.getMergeProgress());
+            this.supportsIOThrottling = supportsIOThrottling;
+            // probably redundant, but better be explicit
+            if (this.supportsIOThrottling == false) {
+                this.rateLimiter.setMBPerSec(Double.POSITIVE_INFINITY);
+            }
+        }
+
+        boolean runNowOrBacklog() {
+            return ThreadPoolMergeScheduler.this.runNowOrBacklog(this);
+        }
+
+        @Override
+        public int compareTo(MergeTask other) {
+            // sort smaller merges first, so they are executed before larger ones
+            return Long.compare(onGoingMerge.getMerge().estimatedMergeBytes, other.onGoingMerge.getMerge().estimatedMergeBytes);
```

Comment:
Can we assert that `estimatedMergeBytes` is > 0? I think it would always be set at this time, but could be good to be sure.

Reply:
Good point! Pushed [cb95685](https://github.com/elastic/elasticsearch/pull/120869/commits/cb956859b1deb106c7147a433483a5fc8e3293fc)

Reply:
Here's a CI run where this assertion fails: https://gradle-enterprise.elastic.co/s/tkf27dbqrex4a .
I could not reproduce it. It's puzzling because the `ConcurrentMergeScheduler` also uses the `estimatedMergeBytes`, though it doesn't assert that it's >0.
I'll keep an eye on it.

Reply:
What is an empty merge? You mean merging together multiple empty segments or merging no segments to one? I find this a bit hard to reason about.

We can punt on it for sure, perhaps add a todo to revisit?

Reply:
>What is an empty merge? You mean merging together multiple empty segments or merging no segments to one? I find this a bit hard to reason about.

I was thinking merging empty segments, but I couldn't trace it in the code, and CI stacktraces are not telling (to myself at least).

I've pushed a TODO comment [e75c94c](https://github.com/elastic/elasticsearch/pull/120869/commits/e75c94cc2ea8db1a3f8d99f675de94ce46888f20)